composer/algorithms/seq_length_warmup/seq_length_warmup.py

```diff
@@ -64,25 +67,31 @@ def set_batch_sequence_length(batch: Dict[str, torch.Tensor], curr_seq_len: int,
     """
 
     assert isinstance(batch, Mapping)
+    assert 'attention_mask' in batch
 
-    if truncate:
-        for k in batch.keys():
-            batch[k] = batch[k][:, :curr_seq_len]
-    else:
-        # ensure new tensor shape is divisible by curr_seq_len
-        input_ids = batch['input_ids'].view(-1)
-        tensor_len = (input_ids.shape[0] // curr_seq_len) * curr_seq_len
+    if curr_seq_len >= batch['attention_mask'].shape[1]:
+        return batch
 
-        input_ids = input_ids[:tensor_len]
-        input_ids = input_ids.view(-1, curr_seq_len)
-        batch['input_ids'] = input_ids
+    # Truncate, but preserve end-of-sequence tokens
+    if preserve_eos:
+        r_idx = torch.arange(batch['attention_mask'].shape[0])
+        # eos_idx should point to the final token index for each batch sample
+        eos_idx = batch['attention_mask'].sum(1).long() - 1
+        # eos_idx_truncated is the same thing, after truncation is applied
+        eos_idx_truncated = eos_idx.clamp(max=curr_seq_len - 1)
 
-        for k, v in batch.items():
-            if k == 'input_ids':
+        for k in batch.keys():
+            if len(batch[k].shape) < 2:
+                continue
+            eos_value = batch[k][r_idx, eos_idx]
+            batch[k] = batch[k][:, :curr_seq_len].contiguous()
```

Comment:
Do we need the contiguous here? I think it forces memory fragmentation, which is inefficient from a hardware perspective.

Reply:
Unfortunately, it is necessary. The BERT forward pass fails without the contiguous. I have expanded the test code to confirm that the model can do a forward/backward pass on the modified batch -- those tests will fail for BERT models if the contiguous is removed.